finality-aleph/src/new_network/substrate.rs

```diff
@@ -33,19 +50,14 @@ impl<B: Block, H: ExHashT> Network for Arc<NetworkService<B, H>> {
         data: impl Into<Vec<u8>> + Send + Sync + 'static,
         peer_id: PeerId,
         protocol: Cow<'static, str>,
-    ) -> Result<(), NetworkSendError> {
-        if let Ok(sender) = self.notification_sender(peer_id.into(), protocol) {
-            sender
-                .ready()
-                .await
-                .ok()
-                .unwrap()
-                .send(data)
-                .map_err(|_| NetworkSendError::Closed)
-        } else {
-            debug!(target: "aleph-network", "Attempted send to peer we are not connected to.");
-            Err(NetworkSendError::Closed)
-        }
+    ) -> Result<(), SendError> {
+        self.notification_sender(peer_id.into(), protocol)
+            .map_err(|_| SendError::NotConnectedToPeer(peer_id))?
+            .ready()
+            .await
```

Comment:
As you rightly pointed out this is explicitly discouraged by the docs. The simplest solution I can think of is to add some pretty tight timeout here and `tokio::select` on these two, returning a timeout error in the appropriate situation.

Reply:
I will add timeout for now and start working on a proper solution right away

Reply:
I added a timeout for 200 ms and it's still a "one liner"